driver-core/src/main/com/mongodb/internal/dns/DefaultDnsResolver.java

```diff
@@ -0,0 +1,148 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.dns;
+
+import com.mongodb.MongoClientException;
+import com.mongodb.MongoConfigurationException;
+
+import javax.naming.Context;
+import javax.naming.NamingEnumeration;
+import javax.naming.NamingException;
+import javax.naming.directory.Attribute;
+import javax.naming.directory.Attributes;
+import javax.naming.directory.InitialDirContext;
+import java.util.ArrayList;
+import java.util.Hashtable;
+import java.util.List;
+
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+
+/**
+ * Utility class for resolving SRV and TXT records.
+ *
+ * <p>This class should not be considered a part of the public API.</p>
+ */
+public final class DefaultDnsResolver implements DnsResolver {
+
+    // The format of SRV record is
+    // priority weight port target.
+    // e.g.
+    // 0 5 5060 example.com.
+    // The priority and weight are ignored, and we just concatenate the host (after removing the ending '.') and port with a
+    // ':' in between, as expected by ServerAddress
+    // It's required that the srvHost has at least three parts (e.g. foo.bar.baz) and that all of the resolved hosts have a parent
+    // domain equal to the domain of the srvHost.
+    @Override
+    public List<String> resolveHostFromSrvRecords(final String srvHost) {
+        String srvHostDomain = srvHost.substring(srvHost.indexOf('.') + 1);
+        List<String> srvHostDomainParts = asList(srvHostDomain.split("\\."));
+        List<String> hosts = new ArrayList<String>();
+        InitialDirContext dirContext = createDnsDirContext();
+        try {
+            Attributes attributes = dirContext.getAttributes("_mongodb._tcp." + srvHost, new String[]{"SRV"});
+            Attribute attribute = attributes.get("SRV");
+            if (attribute == null) {
+                throw new MongoConfigurationException("No SRV records available for host " + srvHost);
+            }
+            NamingEnumeration<?> srvRecordEnumeration = attribute.getAll();
+            while (srvRecordEnumeration.hasMore()) {
+                String srvRecord = (String) srvRecordEnumeration.next();
+                String[] split = srvRecord.split(" ");
+                String resolvedHost = split[3].endsWith(".") ? split[3].substring(0, split[3].length() - 1) : split[3];
+                String resolvedHostDomain = resolvedHost.substring(resolvedHost.indexOf('.') + 1);
+                if (!sameParentDomain(srvHostDomainParts, resolvedHostDomain)) {
+                    throw new MongoConfigurationException(
+                            format("The SRV host name '%s'resolved to a host '%s 'that is not in a sub-domain of the SRV host.",
+                                    srvHost, resolvedHost));
+                }
+                hosts.add(resolvedHost + ":" + split[2]);
+            }
+
+            if (hosts.isEmpty()) {
```

Comment:
This is the one change: throw instead of return an empty list.  Makes it a bit easier on the caller

Reply:
Ack